core/src/apps/binance/sign_tx.py

```diff
@@ -0,0 +1,65 @@
+from trezor.crypto.curve import secp256k1
+from trezor.crypto.hashlib import sha256
+from trezor.messages import MessageType
+from trezor.messages.BinanceSignedTx import BinanceSignedTx
+from trezor.messages.BinanceTxRequest import BinanceTxRequest
+
+from apps.binance import CURVE, helpers, layout
+from apps.common import paths
+
+
+async def sign_tx(ctx, envelope, keychain):
+    # create transaction message -> sign it -> create signature/pubkey message -> serialize all
+    await paths.validate_path(
+        ctx, helpers.validate_full_path, keychain, envelope.address_n, CURVE
+    )
+
+    node = keychain.derive(envelope.address_n)
+
+    tx_req = BinanceTxRequest()
+
+    msg = await ctx.call(
+        tx_req,
+        MessageType.BinanceCancelMsg,
+        MessageType.BinanceOrderMsg,
+        MessageType.BinanceTransferMsg,
+    )
+
+    msg_json = helpers.produce_json_for_signing(envelope, msg)
+    signature_bytes = generate_content_signature(msg_json.encode(), node.private_key())
+
+    if msg.MESSAGE_WIRE_TYPE == MessageType.BinanceTransferMsg:
+        for txinput in msg.inputs:
+            for coin in txinput.coins:
+                await layout.require_confirm_transfer(
+                    ctx, txinput.address, coin.amount, coin.denom
+                )
+        for output in msg.outputs:
+            for coin in output.coins:
+                await layout.require_confirm_transfer(
+                    ctx, output.address, coin.amount, coin.denom
+                )
+    elif msg.MESSAGE_WIRE_TYPE == MessageType.BinanceOrderMsg:
+        await layout.require_confirm_order_side(ctx, msg.side)
+        await layout.require_confirm_order_address(ctx, msg.sender)
+        await layout.require_confirm_order_details(ctx, msg.quantity, msg.price)
+    elif msg.MESSAGE_WIRE_TYPE == MessageType.BinanceCancelMsg:
+        await layout.require_confirm_cancel(ctx, msg.refid)
+    else:
+        raise ValueError("input message unrecognized, is of type " + type(msg).__name__)
+
+    return BinanceSignedTx(
+        signature=signature_bytes, public_key=node.public_key(), json=msg_json
```

Comment:
note: we probably don't need to return the "json" msg we signed since the wallet/client won't need it.

Reply:
IIRC it's something like, Trezor gets a protobuf-ified JSON, signs the canonicalized version, and the wallet will submit a non-canonicalized JSON to the blockchain daemon, which will canonicalize it and check the signature -- so the canonical JSON is never passed around.
Is that correct?
If so, do you think we should drop the field from the protobuf completely?

Reply:
Yes, that is my understanding. IIRC the ledger-binance sdk integration doesn't expect the canonical JSON either so it should be safe to drop it.